File: backend/src/presentation/api/routes/interactive_confirmation.py
"""Interactive Confirmation API - Human-in-the-Loop確認処理

ユーザーからの確認応答を受け取り、後続処理を実行するAPIエンドポイント
"""

from typing import Dict

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from src.application.usecases.interactive_confirmation_usecase import InteractiveConfirmationUseCase
from src.presentation.api.dependencies import get_interactive_confirmation_usecase

router = APIRouter()


class ConfirmationResponseRequest(BaseModel):
    """確認応答リクエスト"""

    confirmation_id: str
    user_response: str
    user_id: str
    session_id: str
    response_metadata: Dict = None


class ConfirmationResponseResponse(BaseModel):
    """確認応答レスポンス"""

    success: bool
    message: str
    followup_action: Dict
    confirmation_id: str
    timestamp: str


@router.post("/process-confirmation-response", response_model=ConfirmationResponseResponse)
async def process_confirmation_response(
    request: ConfirmationResponseRequest,
    interactive_confirmation_usecase: InteractiveConfirmationUseCase = Depends(get_interactive_confirmation_usecase),
) -> ConfirmationResponseResponse:
    """ユーザーの確認応答を処理する

    Args:
        request: 確認応答リクエスト
        interactive_confirmation_usecase: Interactive Confirmation UseCase

    Returns:
        ConfirmationResponseResponse: 処理結果とフォローアップアクション

    Raises:
        HTTPException: 処理エラー時
    """
    try:
        # UseCaseに処理を委譲
        result = await interactive_confirmation_usecase.process_confirmation_response(
            confirmation_id=request.confirmation_id,
            user_response=request.user_response,
            user_id=request.user_id,
            session_id=request.session_id,
            response_metadata=request.response_metadata,
        )

        if not result.get("success", False):
            raise HTTPException(status_code=400, detail=f"応答処理エラー: {result.get('error', '不明なエラー')}")

        return ConfirmationResponseResponse(
            success=True,
            message=result.get("message", "応答を正常に処理しました"),
            followup_action=result.get("followup_action", {}),
            confirmation_id=request.confirmation_id,
            timestamp=result.get("timestamp", ""),
        )

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"内部サーバーエラー: {str(e)}")

File: frontend/src/libs/api/interactive-confirmation.ts
/**
 * Interactive Confirmation API Client
 * Human-in-the-Loop確認処理のためのAPI関数
 */

import { API_BASE_URL } from '@/config/api'

export interface ConfirmationResponseRequest {
  confirmation_id: string
  user_response: string
  user_id: string
  session_id: string
  response_metadata?: Record<string, any>
}

export interface ConfirmationResponseResponse {
  success: boolean
  message: string
  followup_action: Record<string, any>
  confirmation_id: string
  timestamp: string
}

/**
 * ユーザーの確認応答を送信する
 */
export async function sendConfirmationResponse(
  request: ConfirmationResponseRequest
): Promise<ConfirmationResponseResponse> {
  try {
    console.log('📤 送信する確認応答:', request)

    const response = await fetch(`${API_BASE_URL}/api/streaming/process-confirmation-response`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    })

    if (!response.ok) {
      const errorText = await response.text()
      throw new Error(`確認応答送信に失敗しました: ${response.status} ${errorText}`)
    }

    const data = await response.json()
    console.log('📥 確認応答レスポンス:', data)

    return data
  } catch (error) {
    console.error('❌ 確認応答送信エラー:', error)
    throw error
  }
}

/**
 * レスポンスからインタラクティブ確認データを抽出する
 */
export interface InteractiveConfirmationData {
  type: 'interactive_confirmation'
  confirmation_id: string
  question: string
  options: string[]
  context_data?: any
  timeout_seconds?: number
}

export function parseInteractiveConfirmation(response: string): InteractiveConfirmationData | null {
  try {
    // JSON形式の確認データを検出
    const jsonMatch = response.match(/\{[\s\S]*?"type"\s*:\s*"interactive_confirmation"[\s\S]*?\}/);
    if (jsonMatch) {
      const data = JSON.parse(jsonMatch[0]);
      if (data.type === 'interactive_confirmation') {
        return data;
      }
    }

    // requires_user_response フラグを含む応答を検出
    const responseFlagMatch = response.match(/\{[\s\S]*?"requires_user_response"\s*:\s*true[\s\S]*?\}/);
    if (responseFlagMatch) {
      const data = JSON.parse(responseFlagMatch[0]);
      if (data.requires_user_response) {
        return {
          type: 'interactive_confirmation',
          confirmation_id: data.confirmation_data?.confirmation_id || `confirm_${Date.now()}`,
          question: data.confirmation_data?.question || data.message || '確認をお願いします',
          options: data.confirmation_data?.options || ['はい', 'いいえ'],
          context_data: data.context_data || data.confirmation_data?.context_data,
          timeout_seconds: data.confirmation_data?.timeout_seconds || 300
        };
      }
    }

    return null;
  } catch (error) {
    console.error('Interactive confirmation parsing error:', error);
    return null;
  }
}

/**
 * メッセージ内容が確認を必要とするかチェック
 */
export function requiresConfirmation(content: string): boolean {
  // 確認が必要なパターンを検出
  const confirmationPatterns = [
    'requires_user_response',
    'confirmation_id',
    '確認が必要です',
    '登録しますか',
    'interactive_confirmation'
  ]

  return confirmationPatterns.some(pattern => content.includes(pattern))
}